File: doc/codes/python/examples/tutorial/bank.py
import asyncio

import numpy as np

import autd3
import autd3_link_ethercrab as ethercrab
import autd3_pattern as pattern
from autd3.units import m, s


async def main() -> None:
    geometry = autd3.geometry.Geometry([autd3.geometry.Autd3([0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0])])

    client = await autd3.Client.open(
        geometry,
        ethercrab.EtherCrabLinkOption(),
        autd3.ClientConfig(),
    )

    wavelength = pattern.wavelength(340 * m / s)

    # ANCHOR: switch
    # Write focus A to bank B0 and play it.
    target_a = geometry.center() + np.array([0.0, 0.0, 150.0])
    pat_a = geometry.pattern_buffer()
    pattern.focus(
        geometry,
        target_a,
        wavelength,
        pattern.FocusOption(),
        pat_a,
    )
    builder = client.datagram_builder()
    builder.push(autd3.commands.Pattern(pat_a, bank=autd3.value.PatternBank.B0))
    for frame in builder.build():
        await client.send_checked(frame)

    # Write focus B to bank B1, which is not currently playing, then switch to B1.
    # B0 keeps playing cleanly while B1 is being written (double buffering).
    target_b = geometry.center() + np.array([0.0, 30.0, 150.0])
    pat_b = geometry.pattern_buffer()
    pattern.focus(
        geometry,
        target_b,
        wavelength,
        pattern.FocusOption(),
        pat_b,
    )
    builder = client.datagram_builder()
    builder.push(autd3.commands.Pattern(pat_b, bank=autd3.value.PatternBank.B1))
    for frame in builder.build():
        await client.send_checked(frame)
    # ANCHOR_END: switch

    await client.close()


if __name__ == "__main__":
    asyncio.run(main())

File: doc/codes/rust/examples/tutorial/bank.rs
use anyhow::Result;

use autd3_rs::commands::Pattern;
use autd3_rs::geometry::{Autd3, Geometry, offset};
use autd3_rs::units::{m, mm, s};
use autd3_rs::value::PatternBank;
use autd3_rs::{Client, ClientConfig};
use autd3_rs_link_nop::Nop;

#[tokio::main(flavor = "multi_thread")]
async fn main() -> Result<()> {
    let geometry = Geometry::new(vec![Autd3::default()]);

    let client = Client::open(&geometry, Nop, ClientConfig::default()).await?;

    let wavelength = autd3_rs_pattern::wavelength(340.0 * m / s);

    // ANCHOR: switch
    // Write focus A to bank B0 and play it.
    let target_a = geometry.center() + offset(0.0 * mm, 0.0 * mm, 150.0 * mm);
    let mut pat_a = geometry.pattern_buffer();
    autd3_rs_pattern::focus(
        &geometry,
        target_a,
        wavelength,
        &autd3_rs_pattern::FocusOption::default(),
        &mut pat_a,
    );
    let mut builder = client.datagram_builder();
    builder.push(Pattern::with_bank(PatternBank::B0, &pat_a));
    for frame in &builder.build()? {
        client.send_checked(frame).await?;
    }

    // Write focus B to bank B1, which is not currently playing, then switch to B1.
    // B0 keeps playing cleanly while B1 is being written (double buffering).
    let target_b = geometry.center() + offset(0.0 * mm, 30.0 * mm, 150.0 * mm);
    let mut pat_b = geometry.pattern_buffer();
    autd3_rs_pattern::focus(
        &geometry,
        target_b,
        wavelength,
        &autd3_rs_pattern::FocusOption::default(),
        &mut pat_b,
    );
    let mut builder = client.datagram_builder();
    builder.push(Pattern::with_bank(PatternBank::B1, &pat_b));
    for frame in &builder.build()? {
        client.send_checked(frame).await?;
    }
    // ANCHOR_END: switch

    client.close().await?;
    Ok(())
}
